iceoryx_dds/include/iceoryx_dds/dds/cyclone_data_reader.hpp

```diff
@@ -33,28 +33,68 @@ namespace dds
 class CycloneDataReader : public DataReader
 {
   public:
+    /// @brief delete default ctor
     CycloneDataReader() = delete;
+
+    /// @brief Constructor to set cyclone data reader object
+    /// @param[in] serviceId ID of the service
+    /// @param[in] instanceId ID of the instance of the service
+    /// @param[in] eventId ID of the event
     CycloneDataReader(const capro::IdString_t serviceId,
                       const capro::IdString_t instanceId,
                       const capro::IdString_t eventId) noexcept;
+
+    /// @brief Virtual destructor
     virtual ~CycloneDataReader();
 
+    /// @brief copy ctor is explicitly deleted
     CycloneDataReader(const CycloneDataReader&) = delete;
+
+    /// @brief copy assignment is explicitly deleted
     CycloneDataReader& operator=(const CycloneDataReader&) = delete;
+
+    /// @brief move ctor is explicitly deleted
     CycloneDataReader(CycloneDataReader&&) = delete;
+
+    /// @brief move assignment is explicitly deleted
     CycloneDataReader& operator=(CycloneDataReader&&) = delete;
 
+    /// @brief Connect cylcone data reader to the underlying DDS network
     void connect() noexcept override;
 
+    /// @brief Get the size of the next sample if one is available
+    /// @return an optional pointer to the size of the next sample if one is available, cxx::nullopt_t if value is not
+    /// initialized
```

Comment:
If we document overridden methods isn't this duplication ?

Reply:
Good point. So, what's our strategy here?

Reply:
@dkroenke

Reply:
@ithier @marthtz
I would advise to do the following:
The virtual methods in a base class need to be fully documented. Example here: https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/include/iceoryx_posh/roudi/memory/memory_block.hpp

For the overridden methods you can reference to the base class method, but only if that is fully documented. It should be explicitly mentioned that this is an implementation of an virtual method.
https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/include/iceoryx_posh/internal/roudi/memory/mempool_collection_memory_block.hpp

With that, doxygen generates a link to the base class method:
![image](https://user-images.githubusercontent.com/38155883/112534247-208d7e80-8dab-11eb-8734-dd7cb25e0452.png)

Reply:
Agreed, less duplication means less potential errors. However, do you want that applied to the `@brief` section only? So, the implementation should say `@brief Implementation of ` and the virtual one a full description. But what about `@param` and `@return`? Should they stay the same/duplicated or omitted in the implementation?
The example you referenced is slightly unclear/-clean for the return type (e.g. why `type T`, it's not a template?!):
```cpp
    /// @return the required memory as multiple of the alignment
    virtual uint64_t size() const noexcept = 0;
``` 
vs.
```cpp
    /// @return the size of type T
    uint64_t size() const noexcept override;
```

Reply:
@dkroenke @ithier I would partially disagree. Isn't the idea of an overridden method that it can behave differently than the original one?! And when it behaves differently shouldn't it be documented?

In the example from @marthtz it is not required but think of policies for instances.

```cpp
struct LockPolicy {
    virtual void lock() = 0;
    virtual void unlock() = 0;
};

struct ThreadSafe : public LockPolicy {
    /// @brief locks a mutex
    void lock() override;
    /// @brief unlocks a mutex
    void unlock() override;
};

struct SingleThreaded : public LockPolicy {
    /// @brief does nothing
    void lock() override;
    /// @brief does nothing
    void unlock() override;
};
```

@dkroenke with your suggestion we would document the abstract base class fully (which can most likely make no statement on all the possible behaviors) and in the specialization, you would refer to the base class and the user then does not know if the lock is blocking or not.

To summarize. The rule should be: use common sense and when the implementation behaves differently for the user it has to be documented.

Reply:
@elfenpiff yep, totally agree with you. It would have been too easy if this could be simplified but again it's a case by case decision!